File: packages/flutter_gemma_mediapipe/ios/Classes/PlatformServiceImpl.swift
import Flutter
import UIKit

class PlatformServiceImpl : NSObject, PlatformService, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var model: InferenceModel?
    private var session: InferenceSession?

    // Multi-session (.task): concurrently-open sessions keyed by sessionId.
    // The singleton `session` above stays the legacy path; these are the
    // openSession() sessions. Generation is serialized in Dart (a Mutex), so
    // at most one streams at a time — the shared event channel stays
    // unambiguous. Guarded by `sessionMapQueue` for thread-safe access.
    private var sessionMap: [Int64: InferenceSession] = [:]
    private let sessionMapQueue = DispatchQueue(label: "flutter_gemma.sessionMap")

    // 0.15.2: embedding migrated to the shared Dart-FFI + LiteRT path
    // (see `lib/core/litert/litert_embedding_model.dart`). The pigeon
    // surface below is preserved for ABI continuity but no longer
    // backs the runtime path — Dart instantiates LitertEmbeddingModel
    // directly and never calls these methods.

    func createModel(
        maxTokens: Int64,
        modelPath: String,
        loraRanks: [Int64]?,
        preferredBackend: PreferredBackend?,
        maxNumImages: Int64?,
        supportAudio: Bool?,
        completion: @escaping (Result<Void, any Error>) -> Void
    ) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                self.model = try InferenceModel(
                    modelPath: modelPath,
                    maxTokens: Int(maxTokens),
                    supportedLoraRanks: loraRanks?.map(Int.init),
                    maxNumImages: Int(maxNumImages ?? 0),
                    preferredBackend: preferredBackend,
                    supportAudio: supportAudio ?? false
                )
                DispatchQueue.main.async {
                    completion(.success(()))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    func closeModel(completion: @escaping (Result<Void, any Error>) -> Void) {
        session = nil
        sessionMapQueue.sync { sessionMap.removeAll() }
        model = nil
        completion(.success(()))
    }

    func createSession(
        temperature: Double,
        randomSeed: Int64,
        topK: Int64,
        topP: Double?,
        loraPath: String?,
        enableVisionModality: Bool?,
        enableAudioModality: Bool?,
        systemInstruction: String?,
        enableThinking: Bool?,
        completion: @escaping (Result<Void, any Error>) -> Void
    ) {
        guard let inference = model?.inference else {
            completion(.failure(PigeonError(code: "Inference model not created", message: nil, details: nil)))
            return
        }

        if enableThinking == true {
            print("[FlutterGemma] Warning: enableThinking=true is not supported on iOS (MediaPipe). " +
                  "Use Android or Desktop with .litertlm models for Gemma 4 thinking mode.")
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let newSession = try InferenceSession(
                    inference: inference,
                    temperature: Float(temperature),
                    randomSeed: Int(randomSeed),
                    topk: Int(topK),
                    topP: topP,
                    loraPath: loraPath,
                    enableVisionModality: enableVisionModality ?? false,
                    enableAudioModality: enableAudioModality ?? false
                )
                DispatchQueue.main.async {
                    self.session = newSession
                    completion(.success(()))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    func closeSession(completion: @escaping (Result<Void, any Error>) -> Void) {
        session = nil
        completion(.success(()))
    }

    func sizeInTokens(prompt: String, completion: @escaping (Result<Int64, any Error>) -> Void) {
        guard let session = session else {
            completion(.failure(PigeonError(code: "Session not created", message: nil, details: nil)))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let tokenCount = try session.sizeInTokens(prompt: prompt)
                DispatchQueue.main.async { completion(.success(Int64(tokenCount))) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func addQueryChunk(prompt: String, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = session else {
            completion(.failure(PigeonError(code: "Session not created", message: nil, details: nil)))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try session.addQueryChunk(prompt: prompt)
                DispatchQueue.main.async { completion(.success(())) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    // Add method for adding image
    func addImage(imageBytes: FlutterStandardTypedData, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = session else {
            completion(.failure(PigeonError(code: "Session not created", message: nil, details: nil)))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                guard let uiImage = UIImage(data: imageBytes.data) else {
                    DispatchQueue.main.async {
                        completion(.failure(PigeonError(code: "Invalid image data", message: "Could not create UIImage from data", details: nil)))
                    }
                    return
                }

                guard let cgImage = uiImage.cgImage else {
                    DispatchQueue.main.async {
                        completion(.failure(PigeonError(code: "Invalid image format", message: "Could not get CGImage from UIImage", details: nil)))
                    }
                    return
                }

                try session.addImage(image: cgImage)

                DispatchQueue.main.async {
                    completion(.success(()))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    // Add audio input (supported since MediaPipe 0.10.33)
    func addAudio(audioBytes: FlutterStandardTypedData, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = session else {
            completion(.failure(PigeonError(code: "Session not created", message: nil, details: nil)))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try session.addAudio(audio: audioBytes.data)
                DispatchQueue.main.async {
                    completion(.success(()))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    func generateResponse(completion: @escaping (Result<String, any Error>) -> Void) {
        guard let session = session else {
            completion(.failure(PigeonError(code: "Session not created", message: nil, details: nil)))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let response = try session.generateResponse()
                DispatchQueue.main.async { completion(.success(response)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    @available(iOS 13.0, *)
    func generateResponseAsync(completion: @escaping (Result<Void, any Error>) -> Void) {
        print("[PLUGIN LOG] generateResponseAsync called")
        guard let session = session, let eventSink = eventSink else {
            print("[PLUGIN LOG] Session or eventSink not created")
            completion(.failure(PigeonError(code: "Session or eventSink not created", message: nil, details: nil)))
            return
        }

        print("[PLUGIN LOG] Session and eventSink available, starting generation")
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                print("[PLUGIN LOG] Getting async stream from session")
                let stream = try session.generateResponseAsync()
                print("[PLUGIN LOG] Got stream, starting Task")
                Task.detached { [weak self] in
                    guard let self = self else {
                        print("[PLUGIN LOG] Self is nil in Task")
                        return
                    }
                    do {
                        print("[PLUGIN LOG] Starting to iterate over stream")
                        var tokenCount = 0
                        for try await token in stream {
                            tokenCount += 1
                            print("[PLUGIN LOG] Got token #\(tokenCount): '\(token)'")
                            DispatchQueue.main.async {
                                print("[PLUGIN LOG] Sending token to Flutter via eventSink")
                                eventSink(["partialResult": token, "done": false])
                                print("[PLUGIN LOG] Token sent to Flutter")
                            }
                        }
                        print("[PLUGIN LOG] Stream finished after \(tokenCount) tokens")
                        DispatchQueue.main.async {
                            print("[PLUGIN LOG] Sending FlutterEndOfEventStream")
                            eventSink(FlutterEndOfEventStream)
                            print("[PLUGIN LOG] FlutterEndOfEventStream sent")
                        }
                    } catch {
                        print("[PLUGIN LOG] Error in stream iteration: \(error)")
                        DispatchQueue.main.async {
                            eventSink(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil))
                        }
                    }
                }
                DispatchQueue.main.async {
                    print("[PLUGIN LOG] Completing with success")
                    completion(.success(()))
                }
            } catch {
                print("[PLUGIN LOG] Error creating stream: \(error)")
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    func stopGeneration(completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = session else {
            completion(.failure(PigeonError(code: "Session not created", message: nil, details: nil)))
            return
        }

        do {
            try session.cancelGeneration()
            completion(.success(()))
        } catch {
            completion(.failure(error))
        }
    }

    // MARK: - Multi-session (.task) — session-scoped twins keyed by sessionId.
    // The legacy singleton methods above are untouched; these address one of N
    // concurrently-open sessions held in `sessionMap`.

    private func requireSession(_ sessionId: Int64) -> InferenceSession? {
        sessionMapQueue.sync { sessionMap[sessionId] }
    }

    func createSessionForId(
        sessionId: Int64,
        temperature: Double,
        randomSeed: Int64,
        topK: Int64,
        topP: Double?,
        loraPath: String?,
        enableVisionModality: Bool?,
        enableAudioModality: Bool?,
        systemInstruction: String?,
        enableThinking: Bool?,
        completion: @escaping (Result<Void, any Error>) -> Void
    ) {
        guard let inference = model?.inference else {
            completion(.failure(PigeonError(code: "Inference model not created", message: nil, details: nil)))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let newSession = try InferenceSession(
                    inference: inference,
                    temperature: Float(temperature),
                    randomSeed: Int(randomSeed),
                    topk: Int(topK),
                    topP: topP,
                    loraPath: loraPath,
                    enableVisionModality: enableVisionModality ?? false,
                    enableAudioModality: enableAudioModality ?? false
                )
                self.sessionMapQueue.sync {
                    try? self.sessionMap[sessionId]?.cancelGeneration()
                    self.sessionMap[sessionId] = newSession
                }
                DispatchQueue.main.async { completion(.success(())) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func closeSessionId(sessionId: Int64, completion: @escaping (Result<Void, any Error>) -> Void) {
        sessionMapQueue.sync { _ = sessionMap.removeValue(forKey: sessionId) }
        completion(.success(()))
    }

    func sizeInTokensForSession(sessionId: Int64, prompt: String, completion: @escaping (Result<Int64, any Error>) -> Void) {
        guard let session = requireSession(sessionId) else {
            completion(.failure(PigeonError(code: "Session \(sessionId) not found", message: nil, details: nil)))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let n = try session.sizeInTokens(prompt: prompt)
                DispatchQueue.main.async { completion(.success(Int64(n))) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func addQueryChunkToSession(sessionId: Int64, prompt: String, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = requireSession(sessionId) else {
            completion(.failure(PigeonError(code: "Session \(sessionId) not found", message: nil, details: nil)))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try session.addQueryChunk(prompt: prompt)
                DispatchQueue.main.async { completion(.success(())) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func addImageToSession(sessionId: Int64, imageBytes: FlutterStandardTypedData, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = requireSession(sessionId) else {
            completion(.failure(PigeonError(code: "Session \(sessionId) not found", message: nil, details: nil)))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                guard let uiImage = UIImage(data: imageBytes.data), let cgImage = uiImage.cgImage else {
                    DispatchQueue.main.async {
                        completion(.failure(PigeonError(code: "Invalid image data", message: nil, details: nil)))
                    }
                    return
                }
                try session.addImage(image: cgImage)
                DispatchQueue.main.async { completion(.success(())) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func addAudioToSession(sessionId: Int64, audioBytes: FlutterStandardTypedData, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = requireSession(sessionId) else {
            completion(.failure(PigeonError(code: "Session \(sessionId) not found", message: nil, details: nil)))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try session.addAudio(audio: audioBytes.data)
                DispatchQueue.main.async { completion(.success(())) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func generateResponseForSession(sessionId: Int64, completion: @escaping (Result<String, any Error>) -> Void) {
        guard let session = requireSession(sessionId) else {
            completion(.failure(PigeonError(code: "Session \(sessionId) not found", message: nil, details: nil)))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let response = try session.generateResponse()
                DispatchQueue.main.async { completion(.success(response)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    @available(iOS 13.0, *)
    func generateResponseAsyncForSession(sessionId: Int64, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = requireSession(sessionId), let eventSink = eventSink else {
            completion(.failure(PigeonError(code: "Session \(sessionId) or eventSink not available", message: nil, details: nil)))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let stream = try session.generateResponseAsync()
                Task.detached {
                    do {
                        for try await token in stream {
                            DispatchQueue.main.async {
                                eventSink(["partialResult": token, "done": false, "sessionId": sessionId])
                            }
                        }
                        // Tagged completion — NOT FlutterEndOfEventStream (which
                        // would close the channel for every other session).
                        DispatchQueue.main.async {
                            eventSink(["partialResult": "", "done": true, "sessionId": sessionId])
                        }
                    } catch {
                        // Surface as a TAGGED DATA event (not FlutterError,
                        // which the EventChannel broadcasts to every session's
                        // listener and from which Dart can't route by id). Dart
                        // demuxes {code: ERROR, sessionId} and closes only this
                        // session, releasing its generation mutex.
                        DispatchQueue.main.async {
                            eventSink(["code": "ERROR", "message": error.localizedDescription, "sessionId": sessionId])
                        }
                    }
                }
                DispatchQueue.main.async { completion(.success(())) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func stopGenerationForSession(sessionId: Int64, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let session = requireSession(sessionId) else {
            completion(.failure(PigeonError(code: "Session \(sessionId) not found", message: nil, details: nil)))
            return
        }
        do {
            try session.cancelGeneration()
            completion(.success(()))
        } catch {
            completion(.failure(error))
        }
    }

    // 0.15.2: embedding pigeon methods dropped from PlatformService.
    // Dart talks to LiteRT C API directly via dart:ffi
    // (lib/core/litert/litert_embedding_model.dart).

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: packages/flutter_gemma_mediapipe/android/src/main/kotlin/dev/flutterberlin/flutter_gemma_mediapipe/PlatformServiceImpl.kt
package dev.flutterberlin.flutter_gemma_mediapipe

import android.content.Context
import android.util.Log

import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.*

import dev.flutterberlin.flutter_gemma_mediapipe.engines.*

internal class PlatformServiceImpl(
  val context: Context
) : PlatformService, EventChannel.StreamHandler {
  companion object {
    private const val TAG = "FlutterGemmaMediaPipePlugin"
  }
  private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
  private var eventSink: EventChannel.EventSink? = null
  private var streamJob: kotlinx.coroutines.Job? = null  // Track stream collection job
  private val engineLock = Any()  // Lock for thread-safe engine access

  // NEW: Use InferenceEngine abstraction instead of InferenceModel
  @Volatile private var engine: InferenceEngine? = null
  @Volatile private var session: InferenceSession? = null

  // Multi-session (.task): concurrently-open sessions keyed by sessionId.
  // The singleton `session` above stays the legacy path; these are the
  // openSession() sessions. Generation is serialized in Dart (a Mutex), so
  // at most one of these streams at a time — the shared event channel stays
  // unambiguous.
  private val sessionMap = mutableMapOf<Long, InferenceSession>()
  private val sessionMapLock = Any()

  // RAG components
  // 0.15.2: embedding implementation moved to Dart (LitertEmbeddingModel
  // via dart:ffi). The pigeon contract below is kept for ABI continuity
  // but the Dart side never calls into it.

  fun cleanup() {
    scope.cancel()
    streamJob?.cancel()
    streamJob = null
    synchronized(engineLock) {
      session?.close()
      session = null
      engine?.close()
      engine = null
    }
    synchronized(sessionMapLock) {
      sessionMap.values.forEach { runCatching { it.close() } }
      sessionMap.clear()
    }
    // 0.15.2: embedding lifetime managed by Dart (LitertEmbeddingModel).
  }

  override fun createModel(
    maxTokens: Long,
    modelPath: String,
    loraRanks: List<Long>?,
    preferredBackend: PreferredBackend?,
    maxNumImages: Long?,
    supportAudio: Boolean?,
    callback: (Result<Unit>) -> Unit
  ) {
    scope.launch {
      try {
        // Build configuration first (before touching state)
        val config = EngineConfig(
          modelPath = modelPath,
          maxTokens = maxTokens.toInt(),
          supportedLoraRanks = loraRanks?.map { it.toInt() },
          preferredBackend = preferredBackend,
          maxNumImages = maxNumImages?.toInt(),
          supportAudio = supportAudio,
        )

        // Create and initialize new engine BEFORE clearing old state
        // This ensures we don't leave state inconsistent on failure
        val newEngine = EngineFactory.createFromModelPath(modelPath, context)
        newEngine.initialize(config)

        // Only now clear old state and swap in new engine (thread-safe)
        synchronized(engineLock) {
          // Cancel stale stream collector before replacing engine
          streamJob?.cancel()
          streamJob = null
          session?.cancelGeneration()
          try {
            session?.close()
          } catch (e: Exception) {
            Log.w(TAG, "Session close during active inference: ${e.message}")
          }
          session = null
          engine?.close()
          engine = newEngine
        }

        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun closeModel(callback: (Result<Unit>) -> Unit) {
    synchronized(engineLock) {
      try {
        session?.cancelGeneration()
        try {
          session?.close()
        } catch (e: Exception) {
          Log.w(TAG, "Session close during active inference: ${e.message}")
        }
        session = null
        synchronized(sessionMapLock) {
          sessionMap.values.forEach { runCatching { it.close() } }
          sessionMap.clear()
        }
        engine?.close()
        engine = null
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun createSession(
    temperature: Double,
    randomSeed: Long,
    topK: Long,
    topP: Double?,
    loraPath: String?,
    enableVisionModality: Boolean?,
    enableAudioModality: Boolean?,
    systemInstruction: String?,
    enableThinking: Boolean?,
    callback: (Result<Unit>) -> Unit
  ) {
    scope.launch {
      try {
        synchronized(engineLock) {
          val currentEngine = engine
            ?: throw IllegalStateException("Inference model is not created")

          val config = SessionConfig(
            temperature = temperature.toFloat(),
            randomSeed = randomSeed.toInt(),
            topK = topK.toInt(),
            topP = topP?.toFloat(),
            loraPath = loraPath,
            enableVisionModality = enableVisionModality,
            enableAudioModality = enableAudioModality,
            systemInstruction = systemInstruction,
            enableThinking = enableThinking ?: false,
          )

          session?.close()
          session = currentEngine.createSession(config)
        }
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun closeSession(callback: (Result<Unit>) -> Unit) {
    synchronized(engineLock) {
      try {
        session?.close()
        session = null
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun sizeInTokens(prompt: String, callback: (Result<Long>) -> Unit) {
    scope.launch {
      try {
        val currentSession = session
          ?: throw IllegalStateException("Session not created")
        val size = currentSession.sizeInTokens(prompt)
        callback(Result.success(size.toLong()))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun addQueryChunk(prompt: String, callback: (Result<Unit>) -> Unit) {
    scope.launch {
      try {
        val currentSession = session
          ?: throw IllegalStateException("Session not created")
        currentSession.addQueryChunk(prompt)
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun addImage(imageBytes: ByteArray, callback: (Result<Unit>) -> Unit) {
    scope.launch {
      try {
        val currentSession = session
          ?: throw IllegalStateException("Session not created")
        currentSession.addImage(imageBytes)
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun addAudio(audioBytes: ByteArray, callback: (Result<Unit>) -> Unit) {
    scope.launch {
      try {
        session?.addAudio(audioBytes) ?: throw IllegalStateException("Session not created")
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun generateResponse(callback: (Result<String>) -> Unit) {
    scope.launch {
      try {
        val currentSession = session
          ?: throw IllegalStateException("Session not created")
        val result = currentSession.generateResponse()
        callback(Result.success(result))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun generateResponseAsync(callback: (Result<Unit>) -> Unit) {
    scope.launch {
      try {
        val currentSession = session
          ?: throw IllegalStateException("Session not created")
        currentSession.generateResponseAsync()
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun stopGeneration(callback: (Result<Unit>) -> Unit) {
    scope.launch {
      try {
        val currentSession = session
          ?: throw IllegalStateException("Session not created")
        currentSession.cancelGeneration()
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  // === Multi-session (.task) — session-scoped twins keyed by sessionId. ===
  // The legacy singleton methods above are untouched; these address one of N
  // concurrently-open sessions held in [sessionMap].

  private fun requireSession(sessionId: Long): InferenceSession =
    synchronized(sessionMapLock) {
      sessionMap[sessionId]
        ?: throw IllegalStateException("Session $sessionId not found")
    }

  override fun createSessionForId(
    sessionId: Long,
    temperature: Double,
    randomSeed: Long,
    topK: Long,
    topP: Double?,
    loraPath: String?,
    enableVisionModality: Boolean?,
    enableAudioModality: Boolean?,
    systemInstruction: String?,
    enableThinking: Boolean?,
    callback: (Result<Unit>) -> Unit
  ) {
    scope.launch {
      try {
        val config = SessionConfig(
          temperature = temperature.toFloat(),
          randomSeed = randomSeed.toInt(),
          topK = topK.toInt(),
          topP = topP?.toFloat(),
          loraPath = loraPath,
          enableVisionModality = enableVisionModality,
          enableAudioModality = enableAudioModality,
          systemInstruction = systemInstruction,
          enableThinking = enableThinking ?: false,
        )
        // Build the session OUTSIDE the map lock — createSession can be slow.
        val currentEngine = synchronized(engineLock) {
          engine ?: throw IllegalStateException("Inference model is not created")
        }
        val newSession = currentEngine.createSession(config)
        synchronized(sessionMapLock) {
          sessionMap[sessionId]?.let { runCatching { it.close() } }
          sessionMap[sessionId] = newSession
        }
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun closeSessionId(sessionId: Long, callback: (Result<Unit>) -> Unit) {
    try {
      synchronized(sessionMapLock) {
        sessionMap.remove(sessionId)?.let { runCatching { it.close() } }
      }
      callback(Result.success(Unit))
    } catch (e: Exception) {
      callback(Result.failure(e))
    }
  }

  override fun sizeInTokensForSession(
    sessionId: Long,
    prompt: String,
    callback: (Result<Long>) -> Unit
  ) {
    scope.launch {
      try {
        callback(Result.success(requireSession(sessionId).sizeInTokens(prompt).toLong()))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun addQueryChunkToSession(
    sessionId: Long,
    prompt: String,
    callback: (Result<Unit>) -> Unit
  ) {
    scope.launch {
      try {
        requireSession(sessionId).addQueryChunk(prompt)
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun addImageToSession(
    sessionId: Long,
    imageBytes: ByteArray,
    callback: (Result<Unit>) -> Unit
  ) {
    scope.launch {
      try {
        requireSession(sessionId).addImage(imageBytes)
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun addAudioToSession(
    sessionId: Long,
    audioBytes: ByteArray,
    callback: (Result<Unit>) -> Unit
  ) {
    scope.launch {
      try {
        requireSession(sessionId).addAudio(audioBytes)
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun generateResponseForSession(
    sessionId: Long,
    callback: (Result<String>) -> Unit
  ) {
    scope.launch {
      try {
        callback(Result.success(requireSession(sessionId).generateResponse()))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun generateResponseAsyncForSession(
    sessionId: Long,
    callback: (Result<Unit>) -> Unit
  ) {
    scope.launch {
      try {
        val s = requireSession(sessionId)
        val mpSession = s as? dev.flutterberlin.flutter_gemma_mediapipe.engines.mediapipe.MediaPipeSession
          ?: throw IllegalStateException(
            "Session $sessionId does not support tagged async streaming")
        // Tag every chunk with sessionId and push over the shared event
        // channel directly (NOT via endOfStream — that would close the channel
        // for other sessions). Dart demuxes by the sessionId key.
        try {
          mpSession.generateResponseAsyncTagged { result, done ->
            val payload = mapOf(
              "partialResult" to result,
              "done" to done,
              "sessionId" to sessionId,
            )
            scope.launch(Dispatchers.Main) { eventSink?.success(payload) }
          }
        } catch (e: Exception) {
          // Surface a generation-time error as a TAGGED DATA event (not an
          // EventChannel error, which would hit every session's listener and
          // drop the sessionId). Dart demuxes it and closes only this session.
          val errPayload = mapOf(
            "code" to "ERROR",
            "message" to (e.message ?: "generation failed"),
            "sessionId" to sessionId,
          )
          scope.launch(Dispatchers.Main) { eventSink?.success(errPayload) }
        }
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun stopGenerationForSession(
    sessionId: Long,
    callback: (Result<Unit>) -> Unit
  ) {
    scope.launch {
      try {
        requireSession(sessionId).cancelGeneration()
        callback(Result.success(Unit))
      } catch (e: Exception) {
        callback(Result.failure(e))
      }
    }
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    // Cancel previous stream collection to prevent orphaned coroutines
    streamJob?.cancel()
    eventSink = events

    synchronized(engineLock) {
      val currentEngine = engine ?: return

      streamJob = scope.launch {
        launch {
          currentEngine.partialResults.collect { (text, done) ->
            val payload = mapOf("partialResult" to text, "done" to done)
            withContext(Dispatchers.Main) {
              events?.success(payload)
              if (done) {
                events?.endOfStream()
              }
            }
          }
        }

        launch {
          currentEngine.errors.collect { error ->
            withContext(Dispatchers.Main) {
              events?.error("ERROR", error.message, null)
            }
          }
        }
      }
    }
  }

  override fun onCancel(arguments: Any?) {
    streamJob?.cancel()
    streamJob = null
    eventSink = null
  }

  // 0.15.2: embedding pigeon methods dropped from PlatformService contract.
  // Dart now talks to LiteRT C API directly via dart:ffi
  // (lib/core/litert/litert_embedding_model.dart).
}
